Panache entity leads to NPE when no database config is present

From https://code.quarkus.io/?g=com.redhat.natwest&a=todo-app&v=1.0.0-SNAPSHOT&b=MAVEN&s=98e.YjV.ARC.XZs.8IG.0D7&cn=code.quarkus.io

And adding 

```
package org.acme.resteasy;

import io.quarkus.panache.common.Sort;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import java.util.List;

/**
 * @author Emmanuel Bernard emmanuel@hibernate.org
 */
@Path("api")
@Produces(value = MediaType.APPLICATION_JSON)
@Consumes(value = MediaType.APPLICATION_JSON)
public class Todo2Resource {
    @GET
    public List<Todo> getAll() {
        return Todo.listAll(Sort.by("order"));
    }
}
```

```
package org.acme.resteasy;

import io.quarkus.hibernate.orm.panache.PanacheEntity;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.validation.constraints.NotBlank;

/**
 * @author Emmanuel Bernard emmanuel@hibernate.org
 */
@Entity
public class Todo extends PanacheEntity {
    @NotBlank @Column(nullable = false)
    public String title;
    public String url;
    public boolean completed;
    @Column(name="ordering")
    public Integer order;
}
```

I get the following NPE

```
java.lang.NullPointerException
	at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.createBaseQuery(CommonPanacheQueryImpl.java:322)
	at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.createQuery(CommonPanacheQueryImpl.java:277)
	at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.list(CommonPanacheQueryImpl.java:228)
	at io.quarkus.hibernate.orm.panache.runtime.PanacheQueryImpl.list(PanacheQueryImpl.java:149)
	at io.quarkus.hibernate.orm.panache.runtime.JpaOperations$JavaJpaOperations.list(JpaOperations.java:32)
	at io.quarkus.hibernate.orm.panache.runtime.JpaOperations$JavaJpaOperations.list(JpaOperations.java:22)
	at io.quarkus.hibernate.orm.panache.common.runtime.AbstractJpaOperations.listAll(AbstractJpaOperations.java:246)
	at io.quarkus.hibernate.orm.panache.runtime.JpaOperations.listAll(JpaOperations.java:268)
	at org.acme.resteasy.Todo.listAll(Todo.java)
	at org.acme.resteasy.Todo2Resource.getAll(Todo2Resource.java:24)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:643)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:507)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:457)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:459)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:419)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:393)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:131)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:37)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:94)
	at io.quarkus.runtime.CleanableExecutor$CleaningRunnable.run(CleanableExecutor.java:231)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Resulted in: org.jboss.resteasy.spi.UnhandledException: java.lang.NullPointerException
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	... 20 more
```

Which gets fixed when I add in `application.properties`

```
# set DB - only mandatory part
quarkus.datasource.db-kind=postgresql
quarkus.datasource.jdbc.url=jdbc:postgresql:quarkus-demo
quarkus.datasource.username=quarkus-demo
quarkus.datasource.password=quarkus-demo

# Hibernate details
quarkus.hibernate-orm.database.generation=drop-and-create
quarkus.hibernate-orm.log.sql=true
# don't do this at home !!!!!!
%prod.quarkus.hibernate-orm.database.generation=drop-and-create
%prod.quarkus.hibernate-orm.sql-load-script=import.sql
```

This is a regression

Just tried various versions of Quarkus and here is what I see:

With Quarkus `1.3.4.Final`, when `quarkus.datasource.db-kind` is missing, then you get the following build-time exception:

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:1.3.4.Final:build (default) on project todo-app: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor#build threw an exception: io.quarkus.deployment.configuration.ConfigurationError: Hibernate extension cannot guess the dialect as no database kind is specified by 'quarkus.datasource.db-kind'
```

When moving to Quarkus `1.7.5.Final`, there is no build time error and the application fails at runtime with:

```
2020-11-17 08:31:23,575 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /api failed, error id: b3f8b602-558a-4a67-a501-7c87f4b682af-1: org.jboss.resteasy.spi.UnhandledException: javax.persistence.PersistenceException: No EntityManager found. Do you have any JPA entities defined?
```

Finally with Quarkus `1.9.2.Final` (and `1.10.0.CR1`) the runtime exceptions: 

```
Caused by: java.lang.NullPointerException
        at io.quarkus.hibernate.orm.panache.common.runtime.CommonPanacheQueryImpl.createBaseQuery(CommonPanacheQueryImpl.java:322)

```

Now, with Quarkus 1.9.x this warning does show up at build time:

```
[WARNING] [io.quarkus.agroal.deployment.AgroalProcessor] The Agroal dependency is present but no JDBC datasources have been defined.
[WARNING] [io.quarkus.hibernate.orm.deployment.HibernateOrmProcessor] Model classes are defined for the default persistence unit but no default datasource found: the default EntityManagerFactory will not be created.
```.

@gsmet I would like to know if this change from build time to runtime error was intentional or not before proceeding any further.

--- a/extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java
+++ b/extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/AbstractJpaOperations.java
@@ -15,7 +15,11 @@ import javax.persistence.Query;
 import javax.transaction.SystemException;
 import javax.transaction.TransactionManager;
 
+import io.agroal.api.AgroalDataSource;
+import io.quarkus.agroal.DataSource;
 import io.quarkus.arc.Arc;
+import io.quarkus.arc.ArcContainer;
+import io.quarkus.arc.InstanceHandle;
 import io.quarkus.hibernate.orm.PersistenceUnit;
 import io.quarkus.hibernate.orm.runtime.PersistenceUnitUtil;
 import io.quarkus.panache.common.Parameters;
@@ -48,13 +52,35 @@ public abstract class AbstractJpaOperations<PanacheQueryType> {
     }
 
     public static EntityManager getEntityManager(String persistentUnitName) {
+        ArcContainer arcContainer = Arc.container();
         if (persistentUnitName == null || PersistenceUnitUtil.isDefaultPersistenceUnit(persistentUnitName)) {
-            return Arc.container().instance(EntityManager.class).get();
+            InstanceHandle<EntityManager> emHandle = arcContainer.instance(EntityManager.class);
+            if (emHandle.isAvailable()) {
+                return emHandle.get();
+            }
+            if (!arcContainer.instance(AgroalDataSource.class).isAvailable()) {
+                throw new IllegalStateException(
+                        "The default datasource has not been properly configured. See https://quarkus.io/guides/datasource#jdbc-datasource for information on how to do that.");
+            }
+            throw new IllegalStateException(
+                    "No entities were found. Did you forget to annotate your Panache Entity classes with '@Entity'?");
         }
 
-        PersistenceUnit.PersistenceUnitLiteral persistenceUnitLiteral = new PersistenceUnit.PersistenceUnitLiteral(
-                persistentUnitName);
-        return Arc.container().instance(EntityManager.class, persistenceUnitLiteral).get();
+        InstanceHandle<EntityManager> emHandle = arcContainer.instance(EntityManager.class,
+                new PersistenceUnit.PersistenceUnitLiteral(persistentUnitName));
+        if (emHandle.isAvailable()) {
+            return emHandle.get();
+        }
+        if (!arcContainer.instance(AgroalDataSource.class,
+                new DataSource.DataSourceLiteral(persistentUnitName)).isAvailable()) {
+            throw new IllegalStateException(
+                    "The named datasource '" + persistentUnitName
+                            + "' has not been properly configured. See https://quarkus.io/guides/datasource#multiple-datasources for information on how to do that.");
+        }
+        throw new IllegalStateException(
+                "No entities were attached to persistence unit '" + persistentUnitName
+                        + "'. Did you forget to annotate your Panache Entity classes with '@Entity' or improperly configure the 'quarkus.hibernate-orm.\" "
+                        + persistentUnitName + "\".packages' property?");
     }
 
     public static EntityManager getEntityManager() {

--- a/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/ErroneousConfigHotReloadTestCase.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/ErroneousConfigHotReloadTestCase.java
@@ -1,0 +1,48 @@
+package io.quarkus.hibernate.orm.panache.deployment.test;
+
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.not;
+
+import java.util.function.Function;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusDevModeTest;
+import io.restassured.RestAssured;
+
+public class ErroneousConfigHotReloadTestCase {
+    @RegisterExtension
+    final static QuarkusDevModeTest TEST = new QuarkusDevModeTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(UnAnnotatedEntity.class, UnAnnotatedEntityResource.class)
+                    .addAsResource("application-commented-out.properties", "application.properties"));
+
+    @Test
+    public void test() {
+        RestAssured.when().get("/unannotatedEntity").then().statusCode(500).body(containsString("default datasource"))
+                .body(not(containsString("NullPointer")));
+
+        TEST.modifyResourceFile("application.properties", new Function<String, String>() {
+            @Override
+            public String apply(String s) {
+                return s.replace("#", "");
+            }
+        });
+
+        RestAssured.when().get("/unannotatedEntity").then().statusCode(500).body(containsString("@Entity"))
+                .body(not(containsString("NullPointer")));
+
+        TEST.modifySourceFile(UnAnnotatedEntity.class, new Function<String, String>() {
+            @Override
+            public String apply(String s) {
+                return s.replace("//", "");
+            }
+        });
+
+        RestAssured.when().get("/unannotatedEntity").then().statusCode(200);
+    }
+
+}

--- a/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/UnAnnotatedEntity.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/UnAnnotatedEntity.java
@@ -1,0 +1,10 @@
+package io.quarkus.hibernate.orm.panache.deployment.test;
+
+import io.quarkus.hibernate.orm.panache.PanacheEntity;
+
+//import javax.persistence.Entity;
+
+//@Entity
+public class UnAnnotatedEntity extends PanacheEntity {
+    public String name;
+}

--- a/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/UnAnnotatedEntityResource.java
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/UnAnnotatedEntityResource.java
@@ -1,0 +1,18 @@
+package io.quarkus.hibernate.orm.panache.deployment.test;
+
+import java.util.List;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+@Path("unannotatedEntity")
+public class UnAnnotatedEntityResource {
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public List<UnAnnotatedEntity> get() {
+        return UnAnnotatedEntity.listAll();
+    }
+}

--- a/extensions/panache/hibernate-orm-panache/deployment/src/test/resources/application-commented-out.properties
+++ b/extensions/panache/hibernate-orm-panache/deployment/src/test/resources/application-commented-out.properties
@@ -1,0 +1,6 @@
+#quarkus.datasource.db-kind=h2
+#quarkus.datasource.jdbc.url=jdbc:h2:mem:test
+#
+#quarkus.hibernate-orm.dialect=org.hibernate.dialect.H2Dialect
+#quarkus.hibernate-orm.log.sql=true
+#quarkus.hibernate-orm.database.generation=drop-and-create

